src/vos/vos_obj.c

```diff
@@ -131,41 +131,80 @@ obj_punch(daos_handle_t coh, struct vos_object *obj, daos_epoch_t epoch,
 int
 vos_obj_punch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 	      uint32_t pm_ver, uint32_t flags, daos_key_t *dkey,
-	      unsigned int akey_nr, daos_key_t *akeys)
+	      unsigned int akey_nr, daos_key_t *akeys,
+	      struct daos_tx_handle *dth, int dti_cos_count,
+	      struct daos_tx_id *dti_cos)
 {
-	struct vos_pool	  *vpool;
-	struct vos_object *obj;
-	int		   rc;
+	struct vos_container	*cont;
+	struct vos_object	*obj;
+	int			 rc = 0;
+	bool			 found = false;
 
 	D_DEBUG(DB_IO, "Punch "DF_UOID", epoch "DF_U64"\n",
 		DP_UOID(oid), epoch);
 
-	/* NB: punch always generate a new incarnation of the object */
-	rc = vos_obj_hold(vos_obj_cache_current(), coh, oid, epoch,
-			  false, DAOS_INTENT_PUNCH, &obj);
+	vos_dth_set(dth);
+	cont = vos_hdl2cont(coh);
+
+again:
+	rc = vos_tx_begin(cont->vc_pool);
 	if (rc != 0)
-		return rc;
+		goto reset;
 
-	vpool = vos_obj2pool(obj);
+	/* Commit the CoS DTXs via the PUNCH PMDK transaction. */
+	if (dti_cos_count > 0)
+		vos_dtx_commit_internal(cont, dti_cos, dti_cos_count);
 
-	rc = vos_tx_begin(vpool);
-	if (rc != 0)
-		goto exit;
+	if (dth != NULL && !daos_is_zero_dti(&dth->dth_conflict)) {
+		rc = vos_dtx_abort_internal(cont, &dth->dth_conflict, 1, true);
+		daos_dti_gen(&dth->dth_conflict, true);
+		if (rc != 0)
+			goto exit;
+	}
 
-	if (dkey) { /* key punch */
-		rc = key_punch(obj, epoch, pm_ver, dkey,
-			       akey_nr, akeys, flags);
-	} else { /* object punch */
-		rc = obj_punch(coh, obj, epoch, flags);
+	/* NB: punch always generate a new incarnation of the object */
+	rc = vos_obj_hold(vos_obj_cache_current(), coh, oid, epoch,
+			  false, DAOS_INTENT_PUNCH, &obj);
+	if (rc == 0) {
+		found = true;
+		if (dkey) /* key punch */
+			rc = key_punch(obj, epoch, pm_ver, dkey,
+				       akey_nr, akeys, flags);
+		else /* object punch */
+			rc = obj_punch(coh, obj, epoch, flags);
 	}
 
-	rc = vos_tx_end(vpool, rc);
+	if (dth != NULL) {
+		if (rc == 0) {
+			rc = vos_dtx_prepared(dth);
+		} else if (rc == -DER_INPROGRESS &&
+			   dth->dth_flags & DTX_F_AOC &&
+			   !daos_is_zero_dti(&dth->dth_conflict)) {
+			/* XXX: The leader modification has been done,
+			 *	but current replica failed to execute
+			 *	the modification because of some uncommitted
+			 *	DTX, then it should be the case that we have
+			 *	ever failed to abort such DTX on this replica.
+			 *	Let's force abort such conflict uncommitted)
+			 *	DTX and retry.
+			 *
+			 *	Abort current PMDK transaction before retry.
+			 */
+			vos_tx_end(cont->vc_pool, rc);
+			goto again;
```

Comment:
The same logic is shared by vos_update_begin(), vos_update_end() and vos_obj_punch(), I suggest we re-organize it to avoid duplicated code.

Also, I think this 'AOC' should be removed along with 'COS' once the new obj/key punch is implemented. In my opinion, DAOS 2PC should only be aware DAOS idempotent operations (obj/key/record punch, update), there won't be any 'conflict' in DAOS 2PC layer. All the 'conflicts' should be handled in upper layer (perhaps still in VOS) by conditional check or distributed transaction.

Reply:
AOC logic is removed in the new push.